### DISCUSSION TITLE
Dependencies in an SBT plugin

### DISCUSSION THREAD

**Discussion by irundaia** [sbt/sbt#6727] [Q&A]
Hi,

I'm writing an SBT plugin at the moment, and I want to use akka-streams under the hood. So I've written the implementation of my task, including some unit tests, which work. To try it out, I locally published the plugin. 
However, as soon as I attempt to run task, I cannot instantiate the `ActorSystem`. It fails with a `ClassNotFoundException` on `akka.event.DefaultLoggingFilter`. 

So, to me, it seems like the akka dependency is not included correctly on the class path while running it through SBT. Note that if I open an SBT shell in the project directory, of a project using my plugin, I can find the akka dependency listed in the list of plugins, which makes me think that I must have overlooked a setting or some such.

Does anyone have any pointers on how to get a dependency working? 

P.S

Relevant part of the build.sbt file of the plugin
```sbt
name := "my-plugin"
organization := "my.organization"
version := "0.0.1"

libraryDependencies ++= Seq(
  "com.typesafe.akka" %% "akka-stream-typed" % "2.6.17"
)
```

Greatly abstracted/simplified version of the implementation of the plugin
```scala
// Imports omitted for brevity
object MyPlugin extends AutoPlugin {
  override def trigger: PluginTrigger = AllRequirements

  object autoImport {
    object PluginKeys {
      val execute = TaskKey[Seq[File]]("execute", "Execute.")
    }
  }

  import autoImport.PluginKeys._

  val basePluginSettings = Seq(
    execute := Def.task {
      implicit val as = ActorSystem()
      val result = Source(Seq(1,2,3)).runReduce(_ + _)
      // Do stuff with the result
    }
  }
}
```

And the relevant part of the dependant project's plugins.sbt file
```sbt
addSbtPlugin("my.organization" % "my-plugin" % "0.0.1")
```

**Comment by adpi2** [comment]
Hey @irundaia, I tried to reproduce your issue but it was not so easy because:
- The provided build.sbt seems incomplete. At least it should contain `sbtPlugin := true`.
- The provided implementation does not compile.
- What the version of sbt?

Also if you test the plugin locally you should probably use a snapshot version `0.0.1-SNAPSHOT` to bypass the caches when resolving.

I finally got an error but different than yours:
```
[error] (execute) com.typesafe.config.ConfigException$Missing: No configuration setting found for key 'akka'
```

So it seems akka is correctly loaded.

**Reply by irundaia** [reply]
Hi @adpi2,

thanks for taking the time to have a look at my question. 
My apologies, I might have oversimplified my code snippets. To prevent the error you got, I had already made a copy of the reference configuration of akka in my resources, to prevent that.

In any case, I’ve been able to resolve my issues by explicitly selecting a class loader for akka to use to find the configuration and additional classes, i.e.

```scala
ActorSystem(???, ???, ActorSystem.getClass.getClassLoader)
```

Are there any resources, that you know of, that explains how SBT manages it class loaders?
